File: LeetCode/Graph/DFS/JS/Leetcode1022-sum-of-root-to-leaf-binary-numbers.js
var sumRootToLeaf = function (root) {
    let dfs = function (root, x) {
        if (!root) return x;
        x = x * 2 + root.val;
        if (!root.left && !root.right) return x;
        let tmp = 0;
        if (root.left) tmp += dfs(root.left, x);
        if (root.right) tmp += dfs(root.right, x);
        return tmp;
    }


    return dfs(root, 0);
};

File: LeetCode/Graph/DFS/Golang/Leetcode1022-sum-of-root-to-leaf-binary-numbers.go
package sbgo

func sumRootToLeaf(root *TreeNode) int {
	var dfs func(root *TreeNode, x int) int
	dfs = func(root *TreeNode, x int) int {
		if root == nil {
			return x
		}
		x = x*2 + root.Val
		if root.Left == nil && root.Right == nil {
			return x
		}
		tmp := 0
		if root.Left != nil {
			tmp += dfs(root.Left, x)
		}
		if root.Right != nil {
			tmp += dfs(root.Right, x)
		}
		return tmp
	}

	return dfs(root, 0)
}
